### PR TITLE
Split export statements in commonjs to fix hoisting

### PR DESCRIPTION
**Pull request by smvv**
Hi @kdy1 I would like to contribute to SWC and this PR could solve issue #3456.

However, I'm new to Rust (but have experience in JS compiler tech), and I would like to request early feedback regarding the PR.

For example, I would like to request feedback regarding:
1. the overall approach of splitting export declarations and moving those to the bottom.
2. what should we do with import statements? Can we freely move those to the top?
3. `todo()` statements left in the code.
4. I see new `_default` var declarations in the updated test files. I think this comes from `crates/swc_ecma_transforms_module/src/common_js.rs` and I've made a change to avoid the extra vars when `ModuleItem::ModuleDecl(ModuleDecl::ExportDefaultExpr(expr))` is used. Should I also look into the other var declarations, or can we postpone this to another issue/PR?
5. When should I use `DUMMY_SP` exactly as `span`?
 

**Description:**

This transformation will re-ordering the export declarations. The goal is to solve the ReferenceError problem when exports are referencing identifiers declared later in the code.

**BREAKING CHANGE:**

I would like to make the PR non-breaking, but it could be a breaking change as of now.

Note: I still want to remove all `dbg!()` and other commented code.

**Related issue (if exists):**
#4567

### CODE REVIEW

**Comment by kdy1** [comment]
> 1. the overall approach of splitting export declarations and moving those to the bottom.

We can store it in a vector and append the content of the vector to the actual statements.

> 2. what should we do with import statements? Can we freely move those to the top?

Not sure. I don't know the specification.

> 3. `todo()` statements left in the code.

This is the exact thing bothering me these days. Technically it's a task of minifier, but we can't know if `exports.foo` in `exports.foo = void 0;` is side-effect-free.

I'm considering some new techniques, but I have to polish it and write a RFC.

> 4. I see new `_default` var declarations in the updated test files. I think this comes from `crates/swc_ecma_transforms_module/src/common_js.rs` and I've made a change to avoid the extra vars when `ModuleItem::ModuleDecl(ModuleDecl::ExportDefaultExpr(expr))` is used. Should I also look into the other var declarations, or can we postpone this to another issue/PR?

It can be postponed.

5. When should I use `DUMMY_SP` exactly as `span`?
 
If the compiler generate a span and in the case it does not come from user's source code, you should use `DUMMY_SP`.



Btw, thank you for the PR!

**Comment by magic-akari** [comment]
@kdy1  I am working on the esbuild style module transformation.

**Comment by kdy1** [comment]
@magic-akari Cool!
Thank you and feel free to open a draft PR if you want.

I typically do it to prevent duplicate works, but not opening is fine anyway

File: crates/swc_ecma_transforms_module/src/hoist.rs [hunk 1]
```diff
@@ -19,10 +20,124 @@ impl CompilerPass for ModuleHoister {
...
+                            expr: Box::new(Expr::Ident(ident)),
+                        }));
+                    extra.push(export);
+                }
+
+                ModuleItem::ModuleDecl(ModuleDecl::ExportNamed(NamedExport {
+                    // TODO specifiers,
+                    src: None,
+                    type_only: false,
+                    asserts: None,
```
Comment (kdy1) [inline crates/swc_ecma_transforms_module/src/hoist.rs:117]: I think this bound is not required?

File: crates/swc_ecma_transforms_module/src/hoist.rs [hunk 2]
```diff
@@ -19,10 +20,124 @@ impl CompilerPass for ModuleHoister {
...
+                })) => {
+                    let expr = match decl.clone() {
+                        DefaultDecl::Class(expr) => Decl::Class(ClassDecl {
+                            class: expr.class,
+                            ident: expr.ident.unwrap_or(private_ident!("_default")),
+                            declare: false,
+                        }),
+                        DefaultDecl::Fn(expr) => Decl::Fn(FnDecl {
+                            function: expr.function,
+                            ident: expr.ident.unwrap_or(private_ident!("_default")),
```
Comment (kdy1) [inline crates/swc_ecma_transforms_module/src/hoist.rs:91]: This invokes `private_ident` regardless of `expr.ident`.

Comment (kdy1) [reply crates/swc_ecma_transforms_module/src/hoist.rs:91]: Please use `unwrap_or_else` instead

File: crates/swc_ecma_transforms_module/src/hoist.rs [hunk 3]
```diff
@@ -19,10 +20,124 @@ impl CompilerPass for ModuleHoister {
...
+                //     need_hoist_up = true;
+                // }
+                ModuleItem::ModuleDecl(ModuleDecl::ExportDefaultDecl(ExportDefaultDecl {
+                    decl,
+                    ..
+                })) => {
+                    let expr = match decl.clone() {
+                        DefaultDecl::Class(expr) => Decl::Class(ClassDecl {
+                            class: expr.class,
+                            ident: expr.ident.unwrap_or(private_ident!("_default")),
```
Comment (kdy1) [inline crates/swc_ecma_transforms_module/src/hoist.rs:86]: Same here

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you, and sorry for the late review

**Comment by kdy1** [comment]
I think this can be closed, thanks to the new module transforms

**Pull request closed without merging**